primitives/state-machine/src/ext.rs

```diff
@@ -474,29 +516,48 @@ where
 		} else {
 			let storage_key = storage_key.as_ref();
 
-			let (root, is_empty, _) = {
-				let delta = self.overlay.committed.children.get(storage_key)
-					.into_iter()
-					.flat_map(|map| map.clone().into_iter().map(|(k, v)| (k, v.value)))
-					.chain(self.overlay.prospective.children.get(storage_key)
-							.into_iter()
-							.flat_map(|map| map.clone().into_iter().map(|(k, v)| (k, v.value))));
-
-				self.backend.child_storage_root(storage_key, delta)
-			};
-
-			if is_empty {
-				self.overlay.set_storage(storage_key.into(), None);
+			if let Some(child_info) = self.overlay.child_info(storage_key).cloned() {
+				let (root, is_empty, _) = {
+					let delta = self.overlay.committed.children.get(storage_key)
+						.into_iter()
+						.flat_map(|(map, _)| map.clone().into_iter().map(|(k, v)| (k, v.value)))
+						.chain(
+							self.overlay.prospective.children.get(storage_key)
+								.into_iter()
+								.flat_map(|(map, _)| map.clone().into_iter().map(|(k, v)| (k, v.value)))
+						);
+
+					self.backend.child_storage_root(storage_key, child_info.as_ref(), delta)
+				};
+
+				let root = root.encode();
+				if is_empty {
+					self.overlay.set_storage(storage_key.into(), None);
+				} else {
+					self.overlay.set_storage(storage_key.into(), Some(root.clone()));
+				}
```

Comment:
Why is this interesting ?
this root hash gets invalid if another value is set.

Reply:
Yes I think it is not working properly, IIRC correctly it was there to be able when calling 'storage_root' to get a storage root that takes account of child storage root. But in fact 'storage_root' should not ('full_storage_root' should be use).
I think a simple test at state-machine level (to avoid storage_cache) like :
```
let a = self.child_storage(storage_key, key);
self.child_root(storage_key);
let b = self.child_storage(storage_key, key);
assert_eq(a, b)
```
will fail (unless a is not set).
This will need fix in another pr (child_root transaction and root also need to be stored for backend).

Reply:
So currently it doesn't do anything right ?

Reply:
I was looking at it, I remember better, it is specific to externalities: 'storage_root' for externalities got the same semantic as 'full_storage_root' for the overlay, and this was used to cache the result of the child trie root, so on call to child root only we will get the right value.
I will try my test and possibly remove the caching from child_trie_root and those line.

Reply:
Ok I wrote a test and it is actually not broken.
The fact is that child storage root is always fetch from the trie backend (never from the changed value in the overlay), so it is ok to get some cache, but it should definitely be rewrite to get unitary cache from each call to child_storage_root (caching root value from trie backend should be more prioritary).

Reply:
maybe we could write a short comment on top of this if

Reply:
Yes, I already forgot once how it works, so it seems more than usefull.